File: trunk/symmetric-core/src/main/java/org/jumpmind/symmetric/service/impl/OutgoingBatchService.java
/*
 * Licensed to JumpMind Inc under one or more contributor 
 * license agreements.  See the NOTICE file distributed
 * with this work for additional information regarding 
 * copyright ownership.  JumpMind Inc licenses this file
 * to you under the GNU Lesser General Public License (the
 * "License"); you may not use this file except in compliance
 * with the License. 
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, see           
 * <http://www.gnu.org/licenses/>.
 * 
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License. 
 */

package org.jumpmind.symmetric.service.impl;

import java.sql.Types;
import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.apache.commons.lang.StringUtils;
import org.jumpmind.db.sql.ISqlRowMapper;
import org.jumpmind.db.sql.ISqlTransaction;
import org.jumpmind.db.sql.Row;
import org.jumpmind.db.sql.mapper.StringMapper;
import org.jumpmind.symmetric.common.Constants;
import org.jumpmind.symmetric.common.ParameterConstants;
import org.jumpmind.symmetric.common.TableConstants;
import org.jumpmind.symmetric.db.ISymmetricDialect;
import org.jumpmind.symmetric.model.Channel;
import org.jumpmind.symmetric.model.Node;
import org.jumpmind.symmetric.model.NodeChannel;
import org.jumpmind.symmetric.model.NodeGroupChannelWindow;
import org.jumpmind.symmetric.model.NodeHost;
import org.jumpmind.symmetric.model.NodeSecurity;
import org.jumpmind.symmetric.model.OutgoingBatch;
import org.jumpmind.symmetric.model.OutgoingBatch.Status;
import org.jumpmind.symmetric.model.OutgoingBatchSummary;
import org.jumpmind.symmetric.model.OutgoingBatches;
import org.jumpmind.symmetric.service.IClusterService;
import org.jumpmind.symmetric.service.IConfigurationService;
import org.jumpmind.symmetric.service.INodeService;
import org.jumpmind.symmetric.service.IOutgoingBatchService;
import org.jumpmind.symmetric.service.IParameterService;
import org.jumpmind.symmetric.service.ISequenceService;
import org.jumpmind.util.AppUtils;
import org.jumpmind.util.FormatUtils;

/**
 * @see IOutgoingBatchService
 */
public class OutgoingBatchService extends AbstractService implements IOutgoingBatchService {

    private INodeService nodeService;

    private IConfigurationService configurationService;

    private ISequenceService sequenceService;
    
    private IClusterService clusterService;

    public OutgoingBatchService(IParameterService parameterService,
            ISymmetricDialect symmetricDialect, INodeService nodeService,
            IConfigurationService configurationService, ISequenceService sequenceService, IClusterService clusterService) {
        super(parameterService, symmetricDialect);
        this.nodeService = nodeService;
        this.configurationService = configurationService;
        this.sequenceService = sequenceService;
        this.clusterService = clusterService;
        setSqlMap(new OutgoingBatchServiceSqlMap(symmetricDialect.getPlatform(),
                createSqlReplacementTokens()));
    }

    public void markAllAsSentForNode(Node node) {
        OutgoingBatches batches = null;
        do {
            batches = getOutgoingBatches(node, true);
            for (OutgoingBatch outgoingBatch : batches.getBatches()) {
                outgoingBatch.setStatus(Status.OK);
                outgoingBatch.setErrorFlag(false);
                updateOutgoingBatch(outgoingBatch);
            }
        } while (batches.getBatches().size() > 0);
    }

    public void updateAbandonedRoutingBatches() {
        sqlTemplate.update(getSql("updateOutgoingBatchesStatusSql"), Status.NE.name(),
                Status.RT.name());
    }

    public void updateOutgoingBatches(List<OutgoingBatch> outgoingBatches) {
        for (OutgoingBatch batch : outgoingBatches) {
            updateOutgoingBatch(batch);
        }
    }

    public void updateOutgoingBatch(OutgoingBatch outgoingBatch) {
        outgoingBatch.setLastUpdatedTime(new Date());
        outgoingBatch.setLastUpdatedHostName(clusterService.getServerId());
        sqlTemplate.update(
                getSql("updateOutgoingBatchSql"),
                new Object[] { outgoingBatch.getStatus().name(),
                        outgoingBatch.isLoadFlag() ? 1 : 0, outgoingBatch.isErrorFlag() ? 1 : 0,
                        outgoingBatch.getByteCount(), outgoingBatch.getExtractCount(),
                        outgoingBatch.getSentCount(), outgoingBatch.getLoadCount(),
                        outgoingBatch.getDataEventCount(), outgoingBatch.getReloadEventCount(),
                        outgoingBatch.getInsertEventCount(), outgoingBatch.getUpdateEventCount(),
                        outgoingBatch.getDeleteEventCount(), outgoingBatch.getOtherEventCount(),
                        outgoingBatch.getIgnoreCount(), outgoingBatch.getRouterMillis(),
                        outgoingBatch.getNetworkMillis(), outgoingBatch.getFilterMillis(),
                        outgoingBatch.getLoadMillis(), outgoingBatch.getExtractMillis(),
                        outgoingBatch.getSqlState(), outgoingBatch.getSqlCode(),
                        FormatUtils.abbreviateForLogging(outgoingBatch.getSqlMessage()),
                        outgoingBatch.getFailedDataId(), outgoingBatch.getLastUpdatedHostName(),
                        outgoingBatch.getLastUpdatedTime(), outgoingBatch.getBatchId(),
                        outgoingBatch.getNodeId() },
                new int[] { Types.CHAR, Types.NUMERIC, Types.NUMERIC, Types.BIGINT, Types.BIGINT,
                        Types.BIGINT, Types.BIGINT, Types.BIGINT, Types.BIGINT, Types.BIGINT,
                        Types.BIGINT, Types.BIGINT, Types.BIGINT, Types.BIGINT, Types.BIGINT,
                        Types.BIGINT, Types.BIGINT, Types.BIGINT, Types.BIGINT, Types.VARCHAR,
                        Types.NUMERIC, Types.VARCHAR, Types.BIGINT, Types.VARCHAR, Types.TIMESTAMP,
                        Types.NUMERIC, Types.VARCHAR });
    }

    public void insertOutgoingBatch(final OutgoingBatch outgoingBatch) {
        ISqlTransaction transaction = null;
        try {
            transaction = sqlTemplate.startSqlTransaction();
            insertOutgoingBatch(transaction, outgoingBatch);
            transaction.commit();
        } finally {
            close(transaction);
        }
    }

    public void insertOutgoingBatch(ISqlTransaction transaction, OutgoingBatch outgoingBatch) {
        outgoingBatch.setLastUpdatedHostName(clusterService.getServerId());

        long batchId = outgoingBatch.getBatchId();
        if (batchId <= 0) {
            batchId = sequenceService.nextVal(transaction,TableConstants.SYM_OUTGOING_BATCH);
        }
        transaction.prepareAndExecute(getSql("insertOutgoingBatchSql"), batchId, outgoingBatch
                .getNodeId(), outgoingBatch.getChannelId(), outgoingBatch.getStatus().name(),
                outgoingBatch.isLoadFlag() ? 1 : 0, outgoingBatch.isCommonFlag() ? 1 : 0,
                outgoingBatch.getReloadEventCount(), outgoingBatch.getOtherEventCount(),
                outgoingBatch.getLastUpdatedHostName());
        outgoingBatch.setBatchId(batchId);
    }

    public OutgoingBatch findOutgoingBatch(long batchId, String nodeId) {
        List<OutgoingBatch> list = (List<OutgoingBatch>) sqlTemplate.query(
                getSql("selectOutgoingBatchPrefixSql", "findOutgoingBatchSql"),
                new OutgoingBatchMapper(true, false), new Object[] { batchId, nodeId }, new int[] {
                        Types.NUMERIC, Types.VARCHAR });
        if (list != null && list.size() > 0) {
            return list.get(0);
        } else {
            return null;
        }
    }

    public int countOutgoingBatchesInError() {
        return sqlTemplate.queryForInt(getSql("countOutgoingBatchesErrorsSql"));
    }
    
    public int countOutgoingBatchesUnsent() {        
        return sqlTemplate.queryForInt(getSql("countOutgoingBatchesUnsentSql"));
    }

    public int countOutgoingBatches(List<String> nodeIds, List<String> channels,
            List<OutgoingBatch.Status> statuses) {
        Map<String, Object> params = new HashMap<String, Object>();
        params.put("NODES", nodeIds);
        params.put("CHANNELS", channels);
        params.put("STATUSES", toStringList(statuses));

        return sqlTemplate.queryForInt(
                getSql("selectCountBatchesPrefixSql", buildBatchWhere(nodeIds, channels, statuses)),
                params);
    }

    public List<OutgoingBatch> listOutgoingBatches(List<String> nodeIds, List<String> channels,
            List<OutgoingBatch.Status> statuses, long startAtBatchId, final int maxRowsToRetrieve,
            boolean ascending) {

        String where = buildBatchWhere(nodeIds, channels, statuses);
        Map<String, Object> params = new HashMap<String, Object>();
        params.put("NODES", nodeIds);
        params.put("CHANNELS", channels);
        params.put("STATUSES", toStringList(statuses));
        String startAtBatchIdSql = null;
        if (startAtBatchId > 0) {
            if (StringUtils.isBlank(where)) {
                where = " where 1=1 ";
            }
            params.put("BATCH_ID", startAtBatchId);
            if (ascending) {
                startAtBatchIdSql = " and batch_id > :BATCH_ID ";
            } else {
                startAtBatchIdSql = " and batch_id < :BATCH_ID ";
            }
        }

        String sql = getSql("selectOutgoingBatchPrefixSql", where, startAtBatchIdSql,
                ascending ? "order by batch_id asc" : " order by batch_id desc");
        return sqlTemplate.query(sql, maxRowsToRetrieve, new OutgoingBatchMapper(true, false),
                params);

    }

    protected List<String> toStringList(List<OutgoingBatch.Status> statuses) {
        List<String> statusStrings = new ArrayList<String>(statuses.size());
        for (Status status : statuses) {
            statusStrings.add(status.name());
        }
        return statusStrings;

    }

    protected boolean containsOnlyStatus(OutgoingBatch.Status status,
            List<OutgoingBatch.Status> statuses) {
        return statuses.size() == 1 && statuses.get(0) == status;
    }
    
    /**
     * Select batches to process. Batches that are NOT in error will be returned
     * first. They will be ordered by batch id as the batches will have already
     * been created by {@link #buildOutgoingBatches(String)} in channel priority
     * order.
     */
    public OutgoingBatches getOutgoingBatches(Node node, boolean includeDisabledChannels) {
        long ts = System.currentTimeMillis();
        final int maxNumberOfBatchesToSelect = parameterService.getInt(
                ParameterConstants.OUTGOING_BATCH_MAX_BATCHES_TO_SELECT, 1000);
        List<OutgoingBatch> list = (List<OutgoingBatch>) sqlTemplate.query(
                getSql("selectOutgoingBatchPrefixSql", "selectOutgoingBatchSql"),
                maxNumberOfBatchesToSelect, new OutgoingBatchMapper(includeDisabledChannels, true),
                new Object[] { node.getNodeId(), OutgoingBatch.Status.NE.name(),
                        OutgoingBatch.Status.QY.name(), OutgoingBatch.Status.SE.name(),
                        OutgoingBatch.Status.LD.name(), OutgoingBatch.Status.ER.name(),
                        OutgoingBatch.Status.IG.name() }, null);

        OutgoingBatches batches = new OutgoingBatches(list);

        List<NodeChannel> channels = configurationService.getNodeChannels(node.getNodeId(), true);
        batches.sortChannels(channels);

        List<OutgoingBatch> keepers = new ArrayList<OutgoingBatch>();

        for (NodeChannel channel : channels) {
            if (parameterService.is(ParameterConstants.DATA_EXTRACTOR_ENABLED)
                    || channel.getChannelId().equals(Constants.CHANNEL_CONFIG)) {
                keepers.addAll(getBatchesForChannelWindows(batches,
                        node,
                        channel,
                        configurationService.getNodeGroupChannelWindows(
                                parameterService.getNodeGroupId(), channel.getChannelId())));
            }
        }
        batches.setBatches(keepers);

        long executeTimeInMs = System.currentTimeMillis() - ts;
        if (executeTimeInMs > Constants.LONG_OPERATION_THRESHOLD) {
            log.warn("{} took {} ms", "selecting batches to extract", executeTimeInMs);
        }

        return batches;
    }
    
    public List<OutgoingBatch> getBatchesForChannelWindows(OutgoingBatches batches, Node targetNode, NodeChannel channel,
            List<NodeGroupChannelWindow> windows) {
        List<OutgoingBatch> keeping = new ArrayList<OutgoingBatch>();
        List<OutgoingBatch> current = batches.getBatches();
        if (current != null && current.size() > 0) {
            if (inTimeWindow(windows, targetNode)) {
                int maxBatchesToSend = channel.getMaxBatchToSend();
                for (OutgoingBatch outgoingBatch : current) {
                    if (channel.getChannelId().equals(outgoingBatch.getChannelId()) && maxBatchesToSend > 0) {
                        keeping.add(outgoingBatch);
                        maxBatchesToSend--;
                    }
                }
            }
        }
        return keeping;
    }

    /**
     * If {@link NodeGroupChannelWindow}s are defined for this channel, then
     * check to see if the time (according to the offset passed in) is within on
     * of the configured windows.
     */
    public boolean inTimeWindow(List<NodeGroupChannelWindow> windows, Node targetNode) {
        if (windows != null && windows.size() > 0) {
            for (NodeGroupChannelWindow window : windows) {
                String timezoneOffset = null;
                List<NodeHost> hosts = nodeService.findNodeHosts(targetNode.getNodeId());
                if (hosts.size() > 0) {
                    timezoneOffset = hosts.get(0).getTimezoneOffset();
                } else {
                    timezoneOffset = AppUtils.getTimezoneOffset();
                }
                if (window.inTimeWindow(timezoneOffset)) {
                    return true;
                }
            }
            return false;
        } else {
            return true;
        }

    }

    public OutgoingBatches getOutgoingBatchRange(String startBatchId, String endBatchId) {
        OutgoingBatches batches = new OutgoingBatches();
        batches.setBatches(sqlTemplate.query(
                getSql("selectOutgoingBatchPrefixSql", "selectOutgoingBatchRangeSql"),
                new OutgoingBatchMapper(true, false), Long.parseLong(startBatchId),
                Long.parseLong(endBatchId)));
        return batches;
    }

    public OutgoingBatches getOutgoingBatchErrors(int maxRows) {
        OutgoingBatches batches = new OutgoingBatches();
        batches.setBatches(sqlTemplate.query(
                getSql("selectOutgoingBatchPrefixSql", "selectOutgoingBatchErrorsSql"), maxRows,
                new OutgoingBatchMapper(true, false), null, null));
        return batches;
    }

    public boolean isInitialLoadComplete(String nodeId) {
        return areAllLoadBatchesComplete(nodeId)
                && !isUnsentDataOnChannelForNode(Constants.CHANNEL_CONFIG, nodeId);
    }

    public boolean areAllLoadBatchesComplete(String nodeId) {

        NodeSecurity security = nodeService.findNodeSecurity(nodeId);
        if (security == null || security.isInitialLoadEnabled()) {
            return false;
        }

        List<String> statuses = (List<String>) sqlTemplate.query(getSql("initialLoadStatusSql"),
                new StringMapper(), nodeId, 1);
        if (statuses == null || statuses.size() == 0) {
            throw new RuntimeException("The initial load has not been started for " + nodeId);
        }

        for (String status : statuses) {
            if (!Status.OK.name().equals(status)) {
                return false;
            }
        }
        return true;
    }

    public boolean isUnsentDataOnChannelForNode(String channelId, String nodeId) {
        int unsentCount = sqlTemplate.queryForInt(getSql("unsentBatchesForNodeIdChannelIdSql"),
                new Object[] { nodeId, channelId });
        if (unsentCount > 0) {
            return true;
        }

        // Do we need to check for unbatched data?
        return false;
    }

    public List<OutgoingBatchSummary> findOutgoingBatchSummary(Status... statuses) {
        Object[] args = new Object[statuses.length];
        StringBuilder inList = new StringBuilder();
        for (int i = 0; i < statuses.length; i++) {
            args[i] = statuses[i].name();
            inList.append("?,");
        }

        String sql = getSql("selectOutgoingBatchSummaryByStatusSql").replace(":STATUS_LIST",
                inList.substring(0, inList.length() - 1));

        return sqlTemplate.query(sql, new OutgoingBatchSummaryMapper(), args);
    }

    class OutgoingBatchSummaryMapper implements ISqlRowMapper<OutgoingBatchSummary> {
        public OutgoingBatchSummary mapRow(Row rs) {
            OutgoingBatchSummary summary = new OutgoingBatchSummary();
            summary.setBatchCount(rs.getInt("batches"));
            summary.setDataCount(rs.getInt("data"));
            summary.setStatus(Status.valueOf(rs.getString("status")));
            summary.setNodeId(rs.getString("node_id"));
            summary.setOldestBatchCreateTime(rs.getDateTime("oldest_batch_time"));
            return summary;
        }
    }

    class OutgoingBatchMapper implements ISqlRowMapper<OutgoingBatch> {

        private boolean includeDisabledChannels = false;
        private boolean limitBasedOnMaxBatchToSend = false;
        private Map<String, Channel> channels;
        private Map<String, Integer> countByChannel;

        public OutgoingBatchMapper(boolean includeDisabledChannels,
                boolean limitBasedOnMaxBatchToSend) {
            this.includeDisabledChannels = includeDisabledChannels;
            this.limitBasedOnMaxBatchToSend = limitBasedOnMaxBatchToSend;
            this.channels = configurationService.getChannels(false);
            this.countByChannel = new HashMap<String, Integer>();
        }

        public OutgoingBatch mapRow(Row rs) {
            String channelId = rs.getString("channel_id");
            Channel channel = channels.get(channelId);
            Integer count = countByChannel.get(channelId);
            if (count == null) {
                count = 0;
            }
            if (channel != null && (includeDisabledChannels || channel.isEnabled())
                    && (!limitBasedOnMaxBatchToSend || count <= channel.getMaxBatchToSend())) {
                count++;
                OutgoingBatch batch = new OutgoingBatch();
                batch.setChannelId(channelId);
                batch.setNodeId(rs.getString("node_id"));
                batch.setStatus(rs.getString("status"));
                batch.setByteCount(rs.getLong("byte_count"));
                batch.setExtractCount(rs.getLong("extract_count"));
                batch.setSentCount(rs.getLong("sent_count"));
                batch.setLoadCount(rs.getLong("load_count"));
                batch.setDataEventCount(rs.getLong("data_event_count"));
                batch.setReloadEventCount(rs.getLong("reload_event_count"));
                batch.setInsertEventCount(rs.getLong("insert_event_count"));
                batch.setUpdateEventCount(rs.getLong("update_event_count"));
                batch.setDeleteEventCount(rs.getLong("delete_event_count"));
                batch.setOtherEventCount(rs.getLong("other_event_count"));
                batch.setIgnoreCount(rs.getLong("ignore_count"));
                batch.setRouterMillis(rs.getLong("router_millis"));
                batch.setNetworkMillis(rs.getLong("network_millis"));
                batch.setFilterMillis(rs.getLong("filter_millis"));
                batch.setLoadMillis(rs.getLong("load_millis"));
                batch.setExtractMillis(rs.getLong("extract_millis"));
                batch.setSqlState(rs.getString("sql_state"));
                batch.setSqlCode(rs.getInt("sql_code"));
                batch.setSqlMessage(rs.getString("sql_message"));
                batch.setFailedDataId(rs.getLong("failed_data_id"));
                batch.setLastUpdatedHostName(rs.getString("last_update_hostname"));
                batch.setLastUpdatedTime(rs.getDateTime("last_update_time"));
                batch.setCreateTime(rs.getDateTime("create_time"));
                batch.setBatchId(rs.getLong("batch_id"));
                batch.setLoadFlag(rs.getBoolean("load_flag"));
                batch.setErrorFlag(rs.getBoolean("error_flag"));
                batch.setCommonFlag(rs.getBoolean("common_flag"));
                return batch;
            } else {
                return null;
            }
        }
    }

}

File: symmetric-client-clib/src/service/OutgoingBatchService.c
/**
 * Licensed to JumpMind Inc under one or more contributor
 * license agreements.  See the NOTICE file distributed
 * with this work for additional information regarding
 * copyright ownership.  JumpMind Inc licenses this file
 * to you under the GNU General Public License, version 3.0 (GPLv3)
 * (the "License"); you may not use this file except in compliance
 * with the License.
 *
 * You should have received a copy of the GNU General Public License,
 * version 3.0 (GPLv3) along with this library; if not, see
 * <http://www.gnu.org/licenses/>.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
#include "service/OutgoingBatchService.h"
#include "common/Log.h"

SymOutgoingBatch * SymOutgoingBatchService_outgoingBatchMapper(SymRow *row) {
    SymOutgoingBatch *batch = SymOutgoingBatch_new(NULL);
    batch->nodeId = row->getStringNew(row, "node_id");
    batch->status = row->getStringNew(row, "status");
    batch->batchId = row->getLong(row, "batch_id");
    batch->channelId = row->getStringNew(row, "channel_id");
    batch->byteCount = row->getLong(row, "byte_count");
    batch->extractCount = row->getLong(row, "extract_count");
    batch->sentCount = row->getLong(row, "sent_count");
    batch->loadCount = row->getLong(row, "load_count");
    batch->dataEventCount = row->getLong(row, "data_event_count");
    batch->reloadEventCount = row->getLong(row, "reload_event_count");
    batch->insertEventCount = row->getLong(row, "insert_event_count");
    batch->updateEventCount = row->getLong(row, "update_event_count");
    batch->deleteEventCount = row->getLong(row, "delete_event_count");
    batch->otherEventCount = row->getLong(row, "other_event_count");
    batch->ignoreCount = row->getLong(row, "ignore_count");
    batch->routerMillis = row->getLong(row, "router_millis");
    batch->networkMillis = row->getLong(row, "network_millis");
    batch->filterMillis = row->getLong(row, "filter_millis");
    batch->loadMillis = row->getLong(row, "load_millis");
    batch->extractMillis = row->getLong(row, "extract_millis");
    batch->sqlState = row->getStringNew(row, "sql_state");
    batch->sqlCode = row->getInt(row, "sql_code");
    batch->sqlMessage = row->getStringNew(row, "sql_message");
    batch->failedDataId = row->getLong(row, "failed_data_id");
    batch->lastUpdatedHostName = row->getStringNew(row, "last_update_hostname");
    batch->lastUpdatedTime = row->getDate(row, "last_update_time");
    batch->createTime = row->getDate(row, "create_time");
    batch->loadFlag = row->getBoolean(row, "load_flag");
    batch->errorFlag = row->getBoolean(row, "error_flag");
    batch->commonFlag = row->getBoolean(row, "common_flag");
    batch->extractJobFlag = row->getBoolean(row, "extract_job_flag");
    batch->loadId = row->getLong(row, "load_id");
    batch->createBy = row->getStringNew(row, "create_by");
    return batch;
}

void SymOutgoingBatchService_insertOutgoingBatch(SymOutgoingBatchService *this, SymOutgoingBatch *batch) {
    if (batch->batchId <= 0) {
        batch->batchId = this->sequenceService->nextVal(this->sequenceService, SYM_SEQUENCE_OUTGOING_BATCH);
    }

    SymStringArray *args = SymStringArray_new(NULL);
    args->addLong(args, batch->batchId)->add(args, batch->nodeId)->add(args, batch->channelId);
    args->add(args, batch->status)->addInt(args, batch->loadId)->addInt(args, batch->extractJobFlag);
    args->addInt(args, batch->loadFlag)->addInt(args, batch->commonFlag)->addLong(args, batch->reloadEventCount);
    args->addLong(args, batch->otherEventCount)->add(args, batch->lastUpdatedHostName)->add(args, batch->createBy);

    SymSqlTemplate *sqlTemplate = this->platform->getSqlTemplate(this->platform);
    int error;
    sqlTemplate->update(sqlTemplate, SYM_SQL_INSERT_OUTGOING_BATCH, args, NULL, &error);

    args->destroy(args);
}

SymOutgoingBatch * SymOutgoingBatchService_findOutgoingBatch(SymOutgoingBatchService *this, long batchId, char *nodeId) {
    SymSqlTemplate *sqlTemplate = this->platform->getSqlTemplate(this->platform);
    SymStringArray *args = SymStringArray_new(NULL);
    args->addLong(args, batchId);
    args->add(args, nodeId);


    SymStringBuilder *sb = SymStringBuilder_newWithString(SYM_SQL_SELECT_OUTGOING_BATCH_PREFIX);
    if (SymStringUtils_isNotBlank(nodeId)) {
        sb->append(sb, SYM_SQL_FIND_OUTGOING_BATCH);
    } else {
        sb->append(sb, SYM_SQL_FIND_OUTGOING_BATCH_BY_ID_ONLY);
    }

    int error;
    SymList *batches = sqlTemplate->query(sqlTemplate, sb->str, args, NULL, &error, (void *) SymOutgoingBatchService_outgoingBatchMapper);

    SymOutgoingBatch *batch = (SymOutgoingBatch *) batches->get(batches, 0);
    sb->destroy(sb);
    args->destroy(args);
    batches->destroy(batches);
    return batch;
}

SymOutgoingBatches * SymOutgoingBatchService_getOutgoingBatches(SymOutgoingBatchService *this, char *nodeId) {
    time_t ts = time(NULL);

    SymStringBuilder *sb = SymStringBuilder_newWithString(SYM_SQL_SELECT_OUTGOING_BATCH_PREFIX);
    sb->append(sb, SYM_SQL_SELECT_OUTGOING_BATCH);

    SymStringArray *args = SymStringArray_new(NULL);
    args->add(args, nodeId)->add(args, SYM_OUTGOING_BATCH_REQUEST)->add(args, SYM_OUTGOING_BATCH_NEW);
    args->add(args, SYM_OUTGOING_BATCH_QUERYING)->add(args, SYM_OUTGOING_BATCH_SENDING)->add(args, SYM_OUTGOING_BATCH_LOADING);
    args->add(args, SYM_OUTGOING_BATCH_ERROR)->add(args, SYM_OUTGOING_BATCH_IGNORED);

    // TODO: sqlTemplate->queryWithLimit with limit on results
    //int maxNumberOfBatchesToSelect = this->parameterService->getInt(this->parameterService, SYM_OUTGOING_BATCH_MAX_BATCHES_TO_SELECT, 1000);
    int error;
    SymSqlTemplate *sqlTemplate = this->platform->getSqlTemplate(this->platform);
    SymList *list = sqlTemplate->query(sqlTemplate, sb->str, args, NULL, &error, (void *) SymOutgoingBatchService_outgoingBatchMapper);

    SymOutgoingBatches *batches = SymOutgoingBatches_newWithList(NULL, list);

    // TODO: sort channels by processing order and errors, and filter batches for only active channels

    // TODO: filter batches with channel window

    time_t executeTimeInMs = (time(NULL) - ts) * 1000;
    if (executeTimeInMs > SYM_LONG_OPERATION_THRESHOLD) {
        SymLog_info("Selecting %d outgoing batch rows for node %s took %ld ms", list->size, nodeId, executeTimeInMs);
    }

    sb->destroy(sb);
    args->destroy(args);
    list->destroy(list);
    return batches;
}

void SymOutgoingBatchService_updateOutgoingBatch(SymOutgoingBatchService *this, SymOutgoingBatch *batch) {
    batch->lastUpdatedTime = SymDate_new();
    batch->lastUpdatedHostName = SymAppUtils_getHostName();

    SymStringArray *args = SymStringArray_new(NULL);
    args->add(args, batch->status)->addInt(args, batch->loadId)->addInt(args, batch->extractJobFlag);
    args->addInt(args, batch->loadFlag)->addInt(args, batch->errorFlag)->addLong(args, batch->byteCount);
    args->addLong(args, batch->extractCount)->addLong(args, batch->sentCount)->addLong(args, batch->loadCount);
    args->addLong(args, batch->dataEventCount)->addLong(args, batch->reloadEventCount)->addLong(args, batch->insertEventCount);
    args->addLong(args, batch->updateEventCount)->addLong(args, batch->deleteEventCount)->addLong(args, batch->otherEventCount);
    args->addLong(args, batch->ignoreCount)->addLong(args, batch->routerMillis)->addLong(args, batch->networkMillis);
    args->addLong(args, batch->filterMillis)->addLong(args, batch->loadMillis)->addLong(args, batch->extractMillis);
    args->add(args, batch->sqlState)->addInt(args, batch->sqlCode)->add(args, batch->sqlMessage);
    args->addLong(args, batch->failedDataId)->add(args, batch->lastUpdatedHostName)->add(args, batch->lastUpdatedTime->dateTimeString);
    args->addLong(args, batch->batchId)->add(args, batch->nodeId);

    SymSqlTemplate *sqlTemplate = this->platform->getSqlTemplate(this->platform);
    int error;
    sqlTemplate->update(sqlTemplate, SYM_SQL_UPDATE_OUTGOING_BATCH, args, NULL, &error);

    args->destroy(args);
}

int SymOutgoingBatchService_countOutgoingBatchesUnsent(SymOutgoingBatchService *this) {
    SymSqlTemplate *sqlTemplate = this->platform->getSqlTemplate(this->platform);
    int error;
    return sqlTemplate->queryForInt(sqlTemplate, SYM_SQL_COUNT_OUTGOING_BATCHES_UNSENT, NULL, NULL, &error);
}

int SymOutgoingBatchService_countOutgoingBatchesInError(SymOutgoingBatchService *this) {
    SymSqlTemplate *sqlTemplate = this->platform->getSqlTemplate(this->platform);
    int error;
    return sqlTemplate->queryForInt(sqlTemplate, SYM_SQL_COUNT_OUTGOING_BATCHES_ERRORS, NULL, NULL, &error);
}

void SymOutgoingBatchService_destroy(SymOutgoingBatchService *this) {
    free(this);
}

SymOutgoingBatchService * SymOutgoingBatchService_new(SymOutgoingBatchService *this, SymDatabasePlatform *platform, SymParameterService *parameterService,
        SymSequenceService *sequenceService) {
    if (this == NULL) {
        this = (SymOutgoingBatchService *) calloc(1, sizeof(SymOutgoingBatchService));
    }
    this->platform = platform;
    this->parameterService = parameterService;
    this->sequenceService = sequenceService;
    this->findOutgoingBatch = (void *) &SymOutgoingBatchService_findOutgoingBatch;
    this->getOutgoingBatches = (void *) &SymOutgoingBatchService_getOutgoingBatches;
    this->insertOutgoingBatch = (void *) &SymOutgoingBatchService_insertOutgoingBatch;
    this->updateOutgoingBatch = (void *) &SymOutgoingBatchService_updateOutgoingBatch;
    this->countOutgoingBatchesUnsent = (void *) &SymOutgoingBatchService_countOutgoingBatchesUnsent;
    this->countOutgoingBatchesInError = (void *) &SymOutgoingBatchService_countOutgoingBatchesInError;
    this->destroy = (void *) &SymOutgoingBatchService_destroy;
    return this;
}
